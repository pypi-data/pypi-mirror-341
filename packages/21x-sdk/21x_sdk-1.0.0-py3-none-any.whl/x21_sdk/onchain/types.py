# Copyright 2025 21X AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from dataclasses import dataclass
from decimal import Decimal
from enum import Enum, StrEnum

from x21_sdk.onchain import util


class OrderType(StrEnum):
    BUY = "Buy"
    SELL = "Sell"


class OrderBookPhase(Enum):
    NONE = 0
    OPEN_FOR_TRADING = 1
    CLOSED_FOR_TRADING = 2
    MANUAL_HALT = 3
    AUTOMATIC_HALT = 4


@dataclass
class Pair:
    base: Decimal = None
    quote: Decimal = None

    def __post_init__(self):
        self.base = util.to_decimal(self.base)
        self.quote = util.to_decimal(self.quote)


@dataclass
class OrderBookOrders:
    buy: list[Pair]
    sell: list[Pair]


@dataclass
class OrderBookOrderCount:
    buy: int
    sell: int


@dataclass
class PairConfiguration:
    base_scale: Decimal
    quote_scale: Decimal
    base_native_scale: Decimal
    quote_native_scale: Decimal
    maker_commission: int
    taker_commission: int
    market_maker_commission: int
    quote_token_address: str
    base_token_address: str
    whitelist_address: str
    commission_wallet: str
    base_token_fallback_wallet: str
    quote_token_fallback_wallet: str
    liquidity_band: int
    version: str

    base_internal_scale: Decimal = None
    quote_internal_scale: Decimal = None
    combined_internal_scale: Decimal = None

    def __post_init__(self):
        self.base_internal_scale = Decimal(self.base_native_scale) / Decimal(self.base_scale)
        self.quote_internal_scale = Decimal(self.quote_native_scale) / Decimal(self.quote_scale)
        self.combined_internal_scale = Decimal(self.base_internal_scale) * Decimal(self.quote_internal_scale)


@dataclass
class ClientInfo:
    id: int
    can_transfer: bool
    can_buy: bool
    can_sell: bool
    can_cancel: bool
    is_market_maker: bool


@dataclass
class PreTradeControlConfig:
    max_value: int
    min_value: int
    price_collar_factor: int
    max_matches: int


@dataclass
class VolatilityManagementConfig:
    static_threshold: int
    dynamic_threshold: int
