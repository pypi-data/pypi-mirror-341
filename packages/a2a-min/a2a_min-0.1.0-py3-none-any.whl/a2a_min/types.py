"""Types for the A2A Min abstraction layer."""

from typing import Any
from pydantic import BaseModel
from a2a_min.base.types import Message, TaskState, Artifact, TextPart


class AgentInvocationResult(BaseModel):
    """Result of an agent invocation."""
    message: Message
    is_complete: bool = True
    requires_input: bool = False
    metadata: dict[str, Any] | None = None
    
    class Config:
        arbitrary_types_allowed = True

    @classmethod
    def agent_msg(cls, message: str):
        return cls(message=Message(
            role="agent",
            parts=[TextPart(text=message)]
        ))


class TaskUpdate(BaseModel):
    """An update from a streaming task."""
    status: TaskState | None = None
    artifact: Artifact | None = None
    is_final: bool = False
    metadata: dict[str, Any] | None = None

    class Config:
        arbitrary_types_allowed = True
