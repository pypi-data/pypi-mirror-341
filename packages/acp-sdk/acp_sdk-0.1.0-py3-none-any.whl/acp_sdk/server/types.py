from typing import Any

from acp_sdk.models import Await, AwaitResume, Message

RunYield = Message | Await | dict[str | Any]
RunYieldResume = AwaitResume | None
