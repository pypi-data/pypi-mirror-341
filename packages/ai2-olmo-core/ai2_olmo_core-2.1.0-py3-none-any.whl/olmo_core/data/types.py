from typing import Type, Union

import numpy as np

from olmo_core.config import StrEnum

NumpyUIntTypes = Union[Type[np.uint8], Type[np.uint16], Type[np.uint32], Type[np.uint64]]


class NumpyDatasetType(StrEnum):
    """
    An enumeration of the different :class:`NumpyDatasetBase` implementations.
    """

    fsl = "fsl"
    """
    Fixed sequenced length ➡️ :class:`NumpyFSLDataset`.
    """

    padded_fsl = "padded_fsl"
    """
    Padded fixed sequence length ➡️ :class:`NumpyPaddedFSLDataset`.
    """

    vsl = "vsl"
    """
    Variable sequenced length ➡️ :class:`NumpyVSLDataset`.
    """


class NumpyDatasetDType(StrEnum):
    uint8 = "uint8"
    uint16 = "uint16"
    uint32 = "uint32"
    uint64 = "uint64"

    def as_np_dtype(
        self,
    ) -> Union[Type[np.uint8], Type[np.uint16], Type[np.uint32], Type[np.uint64]]:
        return getattr(np, str(self))
