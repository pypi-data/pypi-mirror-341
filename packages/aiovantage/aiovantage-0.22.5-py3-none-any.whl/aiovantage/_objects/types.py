"""Common types."""

from dataclasses import dataclass, field
from decimal import Decimal


@dataclass(kw_only=True)
class Parent:
    """Vantage parent type."""

    vid: int
    position: int = field(metadata={"type": "Attribute"})


@dataclass(kw_only=True)
class Array:
    """Vantage array type."""

    @dataclass(kw_only=True)
    class StringData:
        string: str = ""
        size: int = field(metadata={"name": "size", "type": "Attribute"})

    @dataclass(kw_only=True)
    class BytesData:
        bytes: str = ""
        size: int = field(metadata={"name": "size", "type": "Attribute"})

    val: int | Decimal | None = field(default=None, metadata={"name": "val"})
    string: StringData | None = field(default=None, metadata={"name": "string"})
    bytes: BytesData | None = field(default=None, metadata={"name": "bytes"})
    array: "Array | None" = field(default=None, metadata={"name": "array"})
