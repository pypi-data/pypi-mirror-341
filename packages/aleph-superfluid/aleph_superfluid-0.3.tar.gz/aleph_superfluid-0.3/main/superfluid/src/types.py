from typing import Optional, TypedDict, Dict
from enum import Enum

from eth_typing import HexAddress, HexStr

from .utils import normalize_address, is_permissions_clean
from .errors import SFError


class GetFlowParams:

    def __init__(self, super_token: HexAddress, sender: HexAddress, receiver: HexAddress) -> None:
        """
            @param super_token - The token to be flowed
            @param sender - the sender of the flow
            @param receiver - the receiver of the flow
        """
        self.super_token = normalize_address(super_token)
        self.sender = normalize_address(sender)
        self.receiver = normalize_address(receiver)


class Web3FlowInfo(TypedDict):
    timestamp: int
    flowRate: int
    deposit: int
    owedDeposit: int


class GetAccountFlowInfoParams:

    def __init__(self, super_token: HexAddress, account: HexAddress) -> None:
        """
            @param super_token - The token to be flowed
            @param account - the account to get its info
        """
        self.super_token = normalize_address(super_token)
        self.account = normalize_address(account)


class GetFlowOperatorDataParams:

    def __init__(self, super_token: HexAddress, sender: HexAddress, flow_operator: HexAddress) -> None:
        """
            @param super_token - The token to be flowed
            @param sender - the sender of the flow
            @param flow_operator - the spender
        """
        self.super_token = normalize_address(super_token)
        self.sender = normalize_address(sender)
        self.flow_operator = normalize_address(flow_operator)


class Web3FlowOperatorData(TypedDict):
    flowOperatorId: HexStr
    permissions: int
    flowRateAllowance: int


class GetFlowOperatorDataParamsByID:

    def __init__(self, super_token: HexAddress, flow_operator_id: HexStr) -> None:
        """
            @param super_token - The token to be flowed
            @param flow_operator_id - the flow operator id
        """
        self.super_token = normalize_address(super_token)
        self.flow_operator_id = flow_operator_id


class ShouldUseCallAgreement:

    def __init__(self, should_use_call_agreement: Optional[bool] = None) -> None:
        """
            @param should_use_call_agreement - whether or not to use the host contract
        """
        self.should_use_call_agreement = should_use_call_agreement


class UserData:

    def __init__(self, user_data: Optional[HexStr] = None) -> None:
        """
            @param user_data - Extra data provided
        """
        self.user_data = user_data


class ModifyFlowParams(ShouldUseCallAgreement, UserData):

    def __init__(self, receiver: HexAddress, super_token: HexAddress, flow_rate: Optional[int] = None, sender: Optional[HexAddress] = None, user_data: Optional[HexStr] = None, should_use_call_agreement: Optional[bool] = None) -> None:
        """
            @param receiver - receiver of a flow
            @param super_token - The token to be flowed
            @param flow_rate(Optional) - flow rate for the flow
            @param sender(Optional) - sender of the flow
        """
        super().__init__(should_use_call_agreement)
        UserData.__init__(self, user_data)
        self.receiver = normalize_address(receiver)
        self.super_token = normalize_address(super_token)
        self.flow_rate = flow_rate
        self.sender = normalize_address(sender)


class CreateFlowParams(ModifyFlowParams):

    def __init__(self, sender: HexAddress, receiver: HexAddress, super_token: HexAddress, flow_rate: int, user_data: Optional[HexStr] = None,  should_use_call_agreement: Optional[bool] = None) -> None:

        super().__init__(receiver,
                         super_token, sender=sender, flow_rate=flow_rate, user_data=user_data, should_use_call_agreement=should_use_call_agreement)


class UpdateFlowParams(CreateFlowParams):
    pass


class DeleteFlowParams(ModifyFlowParams):

    def __init__(self, sender: HexAddress, receiver: HexAddress, super_token: HexAddress, user_data: Optional[HexStr] = None, should_use_call_agreement: Optional[bool] = None) -> None:
        super().__init__(receiver,
                         super_token, sender=sender, user_data=user_data, should_use_call_agreement=should_use_call_agreement)


class CreateFlowByOperatorParams(CreateFlowParams):
    pass


class UpdateFlowByOperatorParams(CreateFlowByOperatorParams):
    pass


class SuperTokenFlowRateAllowanceParams(UserData):

    def __init__(self, flow_operator: HexAddress, flow_rate_allowance_delta: int, user_data: Optional[HexStr] = None) -> None:
        """
            @param flow_operator - The operator of the flow
            @param flow_rate_allowance_delta - The amount to increase the flow rate allowance by
        """
        super().__init__(user_data)
        self.flow_operator = normalize_address(flow_operator)
        self.flow_rate_allowance_delta = flow_rate_allowance_delta


class FlowRateAllowanceParams(SuperTokenFlowRateAllowanceParams):

    def __init__(self, super_token: HexAddress, flow_operator: HexAddress, flow_rate_allowance_delta: int, user_data: Optional[HexStr] = None) -> None:
        """ 
            @param flow_operator - The operator of the flow
            @param flow_rate_allowance_delta - The amount to increase the flow rate allowance by
            @param super_token - super token
        """
        super().__init__(flow_operator, flow_rate_allowance_delta, user_data)
        self.super_token = normalize_address(super_token)


class SuperTokenUpdateFlowOperatorPermissionsParams(ShouldUseCallAgreement, UserData):

    def __init__(self, flow_operator: HexAddress, permissions: int, flow_rate_allowance: int, user_data: Optional[HexStr] = None, should_use_call_agreement: Optional[bool] = None) -> None:
        """
            @param flow_operator - The operator of the flow
            @param permissions - Number specifying the permission type
            @param flow_rate_allowance - Allowance to an operator
        """
        super().__init__(should_use_call_agreement)
        UserData.__init__(self, user_data)
        self.flow_operator = normalize_address(flow_operator)
        if is_permissions_clean(permissions):
            self.permissions = permissions
        else:
            raise SFError("UNCLEAN_PERMISSIONS",
                          "The desired permissions are unclean")
        if flow_rate_allowance < 0:
            raise SFError("NEGATIVE_FLOW_ALLOWANCE",
                          "No negative flow allowance allowed")
        else:
            self.flow_rate_allowance = flow_rate_allowance


class UpdateFlowOperatorPermissionsParams(SuperTokenUpdateFlowOperatorPermissionsParams):

    def __init__(self, super_token: HexAddress, flow_operator: HexAddress, permissions: int, flow_rate_allowance: int, user_data: Optional[HexStr] = None, should_use_call_agreement: Optional[bool] = None) -> None:
        """
            @param super_token - The token to be flowed
            @param flow_operator - The operator of the flow
            @param permissions - Number specifying the permission type
            @param flow_rate_allowance - Allowance to an operator
        """
        super().__init__(flow_operator, permissions,
                         flow_rate_allowance, user_data, should_use_call_agreement)
        self.super_token = normalize_address(super_token)


class SuperTokenFullControlParams(ShouldUseCallAgreement, UserData):

    def __init__(self, flow_operator: HexAddress, user_data: Optional[HexStr] = None, should_use_call_agreement: Optional[bool] = None) -> None:
        """
            @param flow_operator - The operator of the flow
        """
        super().__init__(should_use_call_agreement)
        UserData.__init__(self, user_data)
        self.flow_operator = normalize_address(flow_operator)


class FullControlParams(SuperTokenFullControlParams):

    def __init__(self, super_token: HexAddress, flow_operator: HexAddress, user_data: Optional[HexStr] = None, should_use_call_agreement: Optional[bool] = None) -> None:
        """
            @param super_token - The token to be flowed
            @param flow_operator - The operator of the flow
        """
        super().__init__(flow_operator, user_data, should_use_call_agreement)
        self.super_token = super_token


class BatchOperationType(Enum):

    UNSUPPORTED = "UNSUPPORTED"  # 0
    ERC20_APPROVE = "ERC20_APPROVE"  # 1
    ERC20_TRANSFER_FROM = "ERC20_TRANSFER_FROM"  # 2
    ERC777_SEND = "ERC777_SEND"  # 3
    ERC20_INCREASE_ALLOWANCE = "ERC20_INCREASE_ALLOWANCE"  # 4
    ERC20_DECREASE_ALLOWANCE = "ERC20_DECREASE_ALLOWANCE"  # 5
    SUPERTOKEN_UPGRADE = "SUPERTOKEN_UPGRADE"  # 101
    SUPERTOKEN_DOWNGRADE = "SUPERTOKEN_DOWNGRADE"  # 102
    SUPERFLUID_CALL_AGREEMENT = "SUPERFLUID_CALL_AGREEMENT"  # 201
    CALL_APP_ACTION = "CALL_APP_ACTION"  # 202


class NETWORK:

    def __init__(self, network: Dict) -> None:
        self.CHAIN_ID = network["chainId"]
        self.RESOLVER = network["contractsV1"].get("resolver")
        self.HOST = network["contractsV1"].get("host")
        self.GOVERNANCE = network["contractsV1"].get("governance")
        self.CFA_V1 = network["contractsV1"].get("cfaV1")
        self.CFA_V1_FORWARDER = network["contractsV1"].get("cfaV1Forwarder")
        self.IDA_V1 = network["contractsV1"].get("idaV1")
        self.SUPER_TOKEN_FACTORY = network["contractsV1"].get(
            "superTokenFactory")
        self.SUPERFLUID_LOADER = network["contractsV1"].get("superfluidLoader")
        self.TOGA = network["contractsV1"].get("toga")
        self.BATCH_LIQUIDATOR = network["contractsV1"].get("batchLiquidator")
        self.FLOW_SCHEDULER = network["contractsV1"].get("flowScheduler")
        self.VESTING_SCHEDULER = network["contractsV1"].get("vestingScheduler")
        self.SUPER_SPREADER = network["contractsV1"].get("superSpreader")
