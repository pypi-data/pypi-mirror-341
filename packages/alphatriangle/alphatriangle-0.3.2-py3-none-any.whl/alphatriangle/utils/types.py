# File: alphatriangle/utils/types.py
from collections import deque
from collections.abc import Mapping

import numpy as np
from typing_extensions import TypedDict


class StateType(TypedDict):
    grid: np.ndarray  # (C, H, W) float32
    other_features: np.ndarray  # (OtherFeatDim,) float32


# Action representation (integer index)
ActionType = int

# Policy target from MCTS (visit counts distribution)
# Mapping from action index to its probability (normalized visit count)
PolicyTargetMapping = Mapping[ActionType, float]

# Experience tuple stored in buffer
# NOW stores the extracted StateType (features) instead of the raw GameState object.
# Kept as Tuple for performance in buffer operations.
# --- CHANGE: Updated comment for n-step returns (used for distributional target) ---
# The third element (float) represents the calculated n-step return (G_t^n)
# starting from the state represented by the first element (StateType).
# This G_t^n is used by the Trainer to construct the target value distribution.
Experience = tuple[StateType, PolicyTargetMapping, float]
# --- END CHANGE ---

# Batch of experiences for training
ExperienceBatch = list[Experience]

# Output type from the neural network's evaluate method
# (Policy Mapping, Value Estimate)
# Kept as Tuple for performance.
# --- CHANGE: Updated comment for expected value ---
# The second element (float) is the EXPECTED value calculated from the
# predicted value distribution (used for MCTS). The Trainer uses the raw logits.
PolicyValueOutput = tuple[Mapping[ActionType, float], float]
# --- END CHANGE ---

# Type alias for the data structure holding collected statistics
# Maps metric name to a deque of (step, value) tuples
# Kept as Dict[Deque] internally in StatsCollectorActor, type alias is sufficient here.
StatsCollectorData = dict[str, deque[tuple[int, float]]]

# --- Pydantic Models for Data Transfer ---
# SelfPlayResult moved to alphatriangle/rl/types.py to resolve circular import


# --- Prioritized Experience Replay Types ---
# TypedDict for the output of the PER buffer's sample method
class PERBatchSample(TypedDict):
    batch: ExperienceBatch
    indices: np.ndarray
    weights: np.ndarray
