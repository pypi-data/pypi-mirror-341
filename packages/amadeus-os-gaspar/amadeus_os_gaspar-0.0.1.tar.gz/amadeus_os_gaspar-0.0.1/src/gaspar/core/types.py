"""
Shared data types for GASPAR system.
"""

from typing import Any, Dict, Optional, List
from datetime import datetime
from pydantic import BaseModel, Field
from enum import Enum

class PrivacyLevel(str, Enum):
    """Privacy level classification."""
    PUBLIC = "PUBLIC"
    INTERNAL = "INTERNAL"
    CONFIDENTIAL = "CONFIDENTIAL"
    RESTRICTED = "RESTRICTED"

class DataConstraint(BaseModel):
    """Data field constraints."""
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    allowed_values: Optional[List[str]] = None
    regex_pattern: Optional[str] = None
    custom_validation: Optional[str] = None

class PrivacyRule(BaseModel):
    """Privacy rule for a data field."""
    field_name: str
    data_type: str
    allowed: bool = True
    privacy_level: PrivacyLevel
    constraints: Optional[DataConstraint] = None
    description: Optional[str] = None
    required_transformations: List[str] = None

class DataSample(BaseModel):
    """Data sample from feed."""
    timestamp: datetime
    source: str
    data: Dict[str, Any] = None
    metadata: Dict[str, Any] = None