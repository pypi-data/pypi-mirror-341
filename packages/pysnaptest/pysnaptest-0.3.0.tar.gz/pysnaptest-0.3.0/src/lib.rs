use std::collections::{BTreeMap, HashMap};
use std::env::VarError;
use std::ops::Deref;
use std::path::PathBuf;
use std::str::{self, FromStr};
use std::sync::{Mutex, MutexGuard};
use std::{env, path::Path};

use csv::ReaderBuilder;

use insta::internals::{Redaction, SnapshotContents};
use insta::Snapshot;
use insta::{rounded_redaction, sorted_redaction};
use once_cell::sync::Lazy;
use pyo3::types::PyAnyMethods;
use pyo3::FromPyObject;
use pyo3::{
    exceptions::PyValueError,
    pyclass, pyfunction, pymethods, pymodule,
    types::{PyModule, PyModuleMethods},
    wrap_pyfunction, Bound, PyAny, PyErr, PyResult,
};

const PYSNAPSHOT_SUFFIX: &str = "pysnap";

static TEST_NAME_COUNTERS: Lazy<Mutex<BTreeMap<String, usize>>> =
    Lazy::new(|| Mutex::new(BTreeMap::new()));

#[derive(Debug)]
struct Description {
    test_file_path: String,
}

impl Description {
    pub fn new(test_file_path: String) -> Self {
        Self { test_file_path }
    }
}

impl From<Description> for String {
    fn from(val: Description) -> Self {
        format!("Test File Path: {}", val.test_file_path)
    }
}

#[derive(Debug)]
struct PytestInfo {
    test_path: String,
    test_name: String,
}

#[derive(Debug)]
enum Error {
    CouldNotSplit(String),
    InvalidEnvVar(VarError),
    NoTestFile,
}

impl From<Error> for PyErr {
    fn from(value: Error) -> Self {
        match value {
            Error::CouldNotSplit(s) => PyValueError::new_err(format!(
                "Expected '::' to be in PYTEST_CURRENT_TEST string ({s})"
            )),
            Error::InvalidEnvVar(ve) => match ve {
                VarError::NotPresent => PyValueError::new_err("PYTEST_CURRENT_TEST is not set"),
                VarError::NotUnicode(os_string) => PyValueError::new_err(format!(
                    "PYTEST_CURRENT_TEST is not a valid unicode string: {os_string:#?}"
                )),
            },
            Error::NoTestFile => PyValueError::new_err("No test file found"),
        }
    }
}

impl PytestInfo {
    pub fn from_env() -> Result<Self, Error> {
        let pytest_str = env::var("PYTEST_CURRENT_TEST").map_err(Error::InvalidEnvVar)?;
        pytest_str.parse()
    }

    pub fn test_path(&self) -> Result<PathBuf, Error> {
        let path = self.test_path_raw();
        if path.exists() {
            Ok(path)
        } else if let Some(filename) = path.file_name() {
            let mut filepath = PathBuf::from("./");
            filepath.push(filename);
            Ok(filepath)
        } else {
            Err(Error::NoTestFile)
        }
    }

    pub fn test_path_raw(&self) -> PathBuf {
        Path::new(&self.test_path).to_path_buf()
    }
}

impl FromStr for PytestInfo {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (test_path, test_name_and_stage) = s
            .split_once("::")
            .ok_or(Error::CouldNotSplit(s.to_string()))?;

        let test_name = test_name_and_stage
            .split_once(" ")
            .map(|(tn, _stage)| tn)
            .unwrap_or(test_name_and_stage);

        Ok(PytestInfo {
            test_name: test_name.to_string(),
            test_path: test_path.to_string(),
        })
    }
}

#[pyclass(frozen)]
#[derive(Debug)]
struct SnapshotInfo {
    snapshot_folder: PathBuf,
    snapshot_name: String,
    relative_test_file_path: Option<String>,
    allow_duplicates: bool,
}

impl TryFrom<PytestInfo> for SnapshotInfo {
    type Error = PyErr;
    fn try_from(value: PytestInfo) -> Result<Self, Self::Error> {
        let test_file_dir = value
            .test_path()?
            .canonicalize()?
            .parent()
            .ok_or_else(|| {
                PyValueError::new_err(format!(
                    "Invalid test_path: {:?}, not yielding a parent directory",
                    value.test_path_raw()
                ))
            })?
            .join("snapshots");

        let test_name = &value.test_name;
        let test_path = value.test_path_raw();
        let file_name = test_path.file_stem().and_then(|s| s.to_str());

        let name = if let Some(f) = file_name {
            format!("{f}_{test_name}")
        } else {
            test_name.to_string()
        };
        Ok(Self {
            snapshot_folder: test_file_dir,
            snapshot_name: name,
            relative_test_file_path: Some(value.test_path()?.to_string_lossy().to_string()),
            allow_duplicates: false,
        })
    }
}

#[pymethods]
impl SnapshotInfo {
    #[staticmethod]
    #[pyo3(signature = (snapshot_path_override = None, snapshot_name_override = None, allow_duplicates = false))]
    fn from_pytest(
        snapshot_path_override: Option<PathBuf>,
        snapshot_name_override: Option<String>,
        allow_duplicates: bool,
    ) -> PyResult<Self> {
        Ok(
            if let (Some(snapshot_folder), Some(snapshot_name)) = (
                snapshot_path_override.clone(),
                snapshot_name_override.clone(),
            ) {
                Self {
                    snapshot_folder,
                    snapshot_name,
                    relative_test_file_path: None,
                    allow_duplicates,
                }
            } else {
                let pytest_info: SnapshotInfo = PytestInfo::from_env()?.try_into()?;
                Self {
                    snapshot_folder: snapshot_path_override.unwrap_or(pytest_info.snapshot_folder),
                    snapshot_name: snapshot_name_override.map_or(pytest_info.snapshot_name, |v| {
                        v.split('-').next().map_or(v.clone(), |s| s.to_string())
                    }),
                    relative_test_file_path: pytest_info.relative_test_file_path,
                    allow_duplicates,
                }
            },
        )
    }

    pub fn snapshot_folder(&self) -> &PathBuf {
        &self.snapshot_folder
    }

    pub fn last_snapshot_name(&self) -> String {
        let test_idx = Self::counters()
            .get(&self.snapshot_name)
            .cloned()
            .unwrap_or(1);
        self.snapshot_name_with_idx(test_idx)
    }

    pub fn next_snapshot_name(&self) -> String {
        let test_idx = Self::counters()
            .get(&self.snapshot_name)
            .cloned()
            .unwrap_or(0)
            + 1;
        self.snapshot_name_with_idx(test_idx)
    }

    pub fn last_snapshot_path(&self) -> PyResult<PathBuf> {
        Ok(self.snapshot_folder.join(format!(
            "pysnaptest__{}@pysnap.snap",
            self.last_snapshot_name()
        )))
    }

    pub fn next_snapshot_path(&self) -> PyResult<PathBuf> {
        Ok(self.snapshot_folder.join(format!(
            "pysnaptest__{}@pysnap.snap",
            self.next_snapshot_name()
        )))
    }
}

impl SnapshotInfo {
    fn counters<'a>() -> MutexGuard<'a, BTreeMap<String, usize>> {
        TEST_NAME_COUNTERS.lock().unwrap_or_else(|x| x.into_inner())
    }

    fn snapshot_name_with_idx(&self, test_idx: usize) -> String {
        if test_idx == 1 {
            self.snapshot_name.to_string()
        } else {
            format!("{}-{}", self.snapshot_name, test_idx)
        }
    }

    fn snapshot_name(&self) -> String {
        let mut c = Self::counters();
        let mut test_idx = c.get(&self.snapshot_name).cloned().unwrap_or(0);
        if !self.allow_duplicates {
            test_idx += 1;
            c.insert(self.snapshot_name.clone(), test_idx);
        }

        self.snapshot_name_with_idx(test_idx)
    }
}

impl TryInto<insta::Settings> for &SnapshotInfo {
    type Error = PyErr;

    fn try_into(self) -> PyResult<insta::Settings> {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path(self.snapshot_folder());
        settings.set_snapshot_suffix(PYSNAPSHOT_SUFFIX);
        if let Some(relative_test_file_path) = &self.relative_test_file_path {
            settings.set_description(Description::new(relative_test_file_path.clone()));
        }
        settings.set_omit_expression(true);
        Ok(settings)
    }
}

#[derive(Debug)]
pub enum RedactionType {
    Sorted,
    Rounded(usize),
    Standard(String),
}

impl<'source> FromPyObject<'source> for RedactionType {
    #[inline]
    fn extract_bound(ob: &Bound<'source, PyAny>) -> PyResult<Self> {
        if ob.is_none() {
            Ok(RedactionType::Sorted)
        } else if let Ok(decimals) = ob.extract::<usize>() {
            Ok(RedactionType::Rounded(decimals))
        } else if let Ok(redaction) = ob.extract::<String>() {
            Ok(RedactionType::Standard(redaction))
        } else {
            Err(PyErr::new::<pyo3::exceptions::PyTypeError, _>(
                "Unable to extract RedactionType",
            ))
        }
    }
}

impl From<RedactionType> for Redaction {
    fn from(value: RedactionType) -> Self {
        match value {
            RedactionType::Sorted => sorted_redaction(),
            RedactionType::Rounded(decimals) => rounded_redaction(decimals),
            RedactionType::Standard(redaction) => redaction.into(),
        }
    }
}

#[pyclass(unsendable)]
#[derive(Debug)]
pub struct PySnapshot(Snapshot);

#[pymethods]
impl PySnapshot {
    #[staticmethod]
    pub fn from_file(p: PathBuf) -> PyResult<Self> {
        Ok(Self(Snapshot::from_file(&p).map_err(|e| {
            PyValueError::new_err(format!("Unable to load snapshot from {p:?}, details: {e}",))
        })?))
    }

    pub fn contents(&self) -> PyResult<Vec<u8>> {
        Ok(match self.0.contents() {
            SnapshotContents::Text(text_snapshot_contents) => {
                text_snapshot_contents.to_string().as_bytes().to_vec()
            }
            SnapshotContents::Binary(items) => items.deref().to_owned(),
        })
    }
}

#[pyfunction]
#[pyo3(signature = (test_info, result, redactions=None))]
fn assert_json_snapshot(
    test_info: &SnapshotInfo,
    result: &Bound<'_, PyAny>,
    redactions: Option<HashMap<String, RedactionType>>,
) -> PyResult<()> {
    let res: serde_json::Value = pythonize::depythonize(result).unwrap();
    let snapshot_name = test_info.snapshot_name();
    let mut settings: insta::Settings = test_info.try_into()?;

    for (selector, redaction) in redactions.unwrap_or_default() {
        settings.add_redaction(selector.as_str(), redaction)
    }

    settings.bind(|| {
        insta::assert_json_snapshot!(snapshot_name, res);
    });
    Ok(())
}

#[pyfunction]
#[pyo3(signature = (test_info, result, redactions=None))]
fn assert_csv_snapshot(
    test_info: &SnapshotInfo,
    result: &str,
    redactions: Option<HashMap<String, RedactionType>>,
) -> PyResult<()> {
    let mut rdr = ReaderBuilder::new().from_reader(result.as_bytes());
    let columns: Vec<Vec<serde_json::Value>> = vec![rdr
        .headers()
        .expect("Expects csv with headers")
        .into_iter()
        .map(|h| h.into())
        .collect()];
    let records = rdr
        .into_deserialize()
        .collect::<Result<Vec<Vec<serde_json::Value>>, _>>()
        .expect("Failed to parse csv records");
    let res: Vec<Vec<serde_json::Value>> = columns.into_iter().chain(records).collect();

    let snapshot_name = test_info.snapshot_name();
    let mut settings: insta::Settings = test_info.try_into()?;

    for (selector, redaction) in redactions.unwrap_or_default() {
        settings.add_redaction(selector.as_str(), redaction)
    }

    settings.bind(|| {
        insta::assert_csv_snapshot!(snapshot_name, res);
    });
    Ok(())
}

#[pyfunction]
fn assert_binary_snapshot(
    test_info: &SnapshotInfo,
    extension: &str,
    result: Vec<u8>,
) -> PyResult<()> {
    let snapshot_name = test_info.snapshot_name();
    let settings: insta::Settings = test_info.try_into()?;
    settings.bind(|| {
        insta::assert_binary_snapshot!(format!("{snapshot_name}.{extension}").as_str(), result);
    });
    Ok(())
}

#[pyfunction]
fn assert_snapshot(test_info: &SnapshotInfo, result: &Bound<'_, PyAny>) -> PyResult<()> {
    let snapshot_name = test_info.snapshot_name();
    let settings: insta::Settings = test_info.try_into()?;
    settings.bind(|| {
        insta::assert_snapshot!(snapshot_name, result);
    });
    Ok(())
}

#[pymodule]
#[pyo3(name = "_pysnaptest")]
fn pysnaptest(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_class::<SnapshotInfo>()?;

    m.add_function(wrap_pyfunction!(assert_snapshot, m)?)?;
    m.add_function(wrap_pyfunction!(assert_binary_snapshot, m)?)?;
    m.add_function(wrap_pyfunction!(assert_json_snapshot, m)?)?;
    m.add_function(wrap_pyfunction!(assert_csv_snapshot, m)?)?;
    m.add_class::<PySnapshot>()?;
    Ok(())
}

#[cfg(test)]
mod tests {

    use crate::{Error, PytestInfo, SnapshotInfo};

    #[test]
    fn test_into_pyinfo_happy_path() {
        let s = "tests/a/b/test_thing.py::test_a (call)";
        let pti: Result<PytestInfo, Error> = s.parse();
        insta::assert_debug_snapshot!(pti)
    }

    #[test]
    fn test_into_pyinfo_no_trailer() {
        let s = "tests/a/b/test_thing.py::test_a";
        let pti: Result<PytestInfo, Error> = s.parse();
        insta::assert_debug_snapshot!(pti)
    }

    #[test]
    fn test_into_pyinfo_failure_case() {
        let s = "tests/a/b/test_thing.py";
        let pti: Result<PytestInfo, Error> = s.parse();
        insta::assert_debug_snapshot!(pti)
    }

    #[test]
    fn test_snapshot_info_overrides_from_pytest() {
        let snapshot_info = SnapshotInfo::from_pytest(
            Some("folder_path_override".into()),
            Some("snapshot_name_override".into()),
            false,
        )
        .unwrap();
        insta::assert_debug_snapshot!(snapshot_info);
        insta::assert_snapshot!(snapshot_info.snapshot_name(), @"snapshot_name_override");
        insta::assert_snapshot!(snapshot_info.last_snapshot_name(), @"snapshot_name_override");
        insta::assert_snapshot!(snapshot_info.next_snapshot_name(), @"snapshot_name_override-2");
        insta::assert_snapshot!(snapshot_info.snapshot_name(), @"snapshot_name_override-2");
        insta::assert_snapshot!(snapshot_info.last_snapshot_name(), @"snapshot_name_override-2");
        insta::assert_snapshot!(snapshot_info.next_snapshot_name(), @"snapshot_name_override-3");
    }
}
