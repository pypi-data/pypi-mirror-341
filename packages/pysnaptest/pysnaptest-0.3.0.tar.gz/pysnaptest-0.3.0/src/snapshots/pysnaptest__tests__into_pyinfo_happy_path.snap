---
source: src/lib.rs
expression: pti
---
Ok(
    PytestInfo {
        test_path: "tests/a/b/test_thing.py",
        test_name: "test_a",
    },
)
